File: tools/remove_bg.py
"""
remove_bg.py — remove background from an image using rembg (on-device AI).

Usage:
    python tools/remove_bg.py <input> <output> [--model u2net|isnet-general-use]

Zero-cost, runs entirely offline. Downloads the model on first use (~200 MB).
"""
import sys
import os
import argparse

def main():
    parser = argparse.ArgumentParser(description="Remove image background via rembg")
    parser.add_argument("input", help="Path to input image")
    parser.add_argument("output", help="Path to output image (PNG recommended)")
    parser.add_argument("--model", default="u2net",
                        choices=["u2net", "u2net_human_seg", "u2netp", "isnet-general-use"],
                        help="Rembg model to use (default: u2net)")
    args = parser.parse_args()

    if not os.path.isfile(args.input):
        print(f"ERROR: input not found: {args.input}", file=sys.stderr)
        sys.exit(1)

    os.makedirs(os.path.dirname(args.output) or ".", exist_ok=True)

    try:
        from rembg import remove, new_session
        session = new_session(args.model)
        with open(args.input, "rb") as f:
            input_data = f.read()
        output_data = remove(input_data, session=session)
        with open(args.output, "wb") as f:
            f.write(output_data)
        print(f"OK: {args.input} -> {args.output}")
    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: src/agentic/operations/remove-bg.ts
/**
 * remove-bg.ts — remove background from an image using on-device AI (rembg).
 *
 * Calls the bundled Python script (tools/remove_bg.py) in the project's venv.
 * Zero API cost, runs entirely offline. Model downloads on first use (~200 MB).
 *
 * Usage:
 *   const r = await removeBackground('input/visuals/photo.jpg', 'input/visuals/photo_nobg.png');
 */

import * as fs from 'fs';
import * as path from 'path';
import { spawn } from 'child_process';

export interface RemoveBgResult {
    ok: boolean;
    output?: string;
    detail: string;
}

const python: string = (() => {
    const p = path.resolve(process.cwd(), 'venv', 'Scripts', 'python.exe');
    if (fs.existsSync(p)) return p;
    // fallback: system python
    return 'python3';
})();

const script: string = (() => {
    const p = path.resolve(process.cwd(), 'tools', 'remove_bg.py');
    return fs.existsSync(p) ? p : '';
})();

/**
 * Remove the background from an image using rembg.
 *
 * @param input   Path to the source image (jpg/png/webp).
 * @param output  Path for the output PNG (transparent background).
 * @param model   Model name: 'u2net' (default, balanced), 'u2netp' (fast/light),
 *                'isnet-general-use' (best quality), 'u2net_human_seg' (people).
 */
export async function removeBackground(
    input: string,
    output: string,
    model: string = 'u2net',
): Promise<RemoveBgResult> {
    if (!fs.existsSync(input)) return { ok: false, detail: `input not found: ${input}` };
    if (!script) return { ok: false, detail: 'tools/remove_bg.py not found' };
    if (!fs.existsSync(python)) return { ok: false, detail: `python not found at ${python}` };

    fs.mkdirSync(path.dirname(output), { recursive: true });

    return new Promise((resolve) => {
        const args = [script, input, output, '--model', model];
        const child = spawn(python, args, { stdio: ['ignore', 'pipe', 'pipe'] });
        let out = '';
        let err = '';

        child.stdout?.on('data', (d: Buffer) => { out += d.toString(); });
        child.stderr?.on('data', (d: Buffer) => { err += d.toString(); });

        child.on('close', (code) => {
            if (code === 0 && fs.existsSync(output) && fs.statSync(output).size > 0) {
                resolve({ ok: true, output, detail: `background removed: ${input} -> ${output}` });
            } else {
                const detail = err.slice(-600) || out.slice(-600) || `exit code ${code}`;
                resolve({ ok: false, detail: `remove-bg failed: ${detail}` });
            }
        });

        child.on('error', (e) => {
            resolve({ ok: false, detail: `remove-bg spawn error: ${e.message}` });
        });
    });
}
